src/ParallelAlgorithms/Interpolation/Protocols/ComputeTargetPoints.hpp

```diff
@@ -0,0 +1,90 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#pragma once
+
+#include <type_traits>
+
+#include "DataStructures/DataBox/DataBox.hpp"
+#include "DataStructures/Tensor/TypeAliases.hpp"
+#include "Utilities/TMPL.hpp"
+
+/// \cond
+namespace Frame {
+struct Grid;
+struct Inertial;
+struct Distorted;
+}  // namespace Frame
+struct DataVector;
+namespace Tags {
+struct Time;
+}  // namespace Tags
+/// \endcond
+
+namespace intrp::protocols {
+/*!
+ * \brief A protocol for the type alias `compute_target_points` found in an
+ * InterpolationTargetTag.
+ *
+ * \details A struct conforming to the `ComputeTargetPoints` protocol must have
+ *
+ * - a type alias `is_sequential` that is either `std::true_type` or
+ *   `std::false_type` which indicates if an interpolations depends on previous
+ *   interpolations' results
+ *
+ * - a type alias `frame` that denotes the frame the target points are computed
+ *   in
+ *
+ * - a function `points` with the following signature
+ *
+ * \snippet Helpers/ParallelAlgorithms/Interpolation/Examples.hpp points
+ *
+ *   that will compute the points in the given `frame`.
+ *
+ * A struct that computes the target points can optionally have any of these
+ * members as well:
+ *
+ * - a type alias `simple_tags` that list simple tags to be added to the DataBox
+ *   of the InterpolationTarget
+ *
+ * - a type alias `compute_tags` that list compute tags to be added to the
+ *   DataBox of the InterpolationTarget
+ *
+ * - a type alias `const_global_cache_tags` with tags to be put in the
+ *   GlobalCache
+ *
+ * - a function `initialize` with the following signature
+ *
+ * \snippet Helpers/ParallelAlgorithms/Interpolation/Examples.hpp initialize
+ *
+ *   that is run during the Initialization phase of the InterpolationTarget and
+ *   can initialize any of the `simple_tags` added.
+ */
+struct ComputeTargetPoints {
+  template <typename ConformingType>
+  struct test {
+    struct DummyMetavariables;
+    struct DummyTag : db::SimpleTag {
+      using type = int;
+    };
+
+    using is_sequential = typename ConformingType::is_sequential;
+    static_assert(std::is_same_v<is_sequential, std::true_type> or
+                  std::is_same_v<is_sequential, std::false_type>);
+
+    using frame = typename ConformingType::frame;
+
+    template <size_t Dim>
+    static constexpr bool conforms = std::is_same_v<
+        tnsr::I<DataVector, Dim, frame>,
+        decltype(ConformingType::points(
+            std::declval<const db::DataBox<DummyTag>&>(),
+            std::declval<const tmpl::type_<DummyMetavariables>&>(),
+            std::declval<const ::Tags::Time&>()))>;
```

Comment:
The function takes a *tag* as argument?

Reply:
I do not remember the exact details, but there are places where the definition of a template is not instantiated if the declaration is sufficient to determine some compile-time value, such as the return type of a function.  (As an extreme example, see `std::declval` itself, which has no definition.)  That's probably why this compiles with bad tags.  (This might be the concept "ODR used", but I'd have to look up the details to be sure.)

Reply:
@wthrowe oops my bad, this shouldn't take a tag, just a double. And yes @wthrowe's explanation is correct. These functions have templates that pick up on all the internal tags and such because those tags can be different for every ComputeTargetPoints like you mentioned @nilsvu. In all of these protocols that have a `static_assert` where I use `std::declval`, I'm checking for 3 things,

1. The conforming type has a member function with this name
2. This member function takes the correct number of arguments and that these arguments are the "correct" type, where "correct" doesn't mean all internals of every type are exactly what they would be when actually used. Just the outermost type is correct (including const, reference, gsl::not_null, etc...). Here, this means that the tags inside the DataBox get picked up by the template so they can be *anything* (like a DummyTag)
3. The return type of this member function is correct.

The fact that this is all compile time, and that we never reach into the internals of this member function in the protocol to check if it conforms, allow us to check its signature without knowing the proper types. So when @wthrowe says this compiles with "bad tags", I don't think that's accurate. It is a **feature** of protocols to be able to check function signatures that have templates and give them fake tags.

I could, in theory, expose enough type aliases everywhere in all of these structs to use the proper tags when checking the function signature. This would constrain the function signatures even more. This will be some extra work though, but is probably doable.

Reply:
By "bad" I just meant something that wouldn't work if you actually tried to use the thing.  I don't have a problem with how the test is being done.